rfcs/0000-cosign-integration.md

```diff
@@ -0,0 +1,143 @@
+
+Moreover, `cosign` is a project that has had a lot of activity and community
+involvement recently.
+
+## Outcome
+
+`kpack` integrates with `cosign` to sign images it builds and push the
+signatures to a registry so that users can ensure the chain of custody of a
+generated artifact.
+
+This proposal aims to cover specifically the flow of signing an image produced
+by `kpack Image` resource without verification of base or builder images pulled
+in the process.
+
+## Actions to take
+
+### Enable image signing using `cosign`
+
+- Implement a flow that generates the signature payload for the image, then
+  calculates its signature and pushes it either to the registry where the image
+  is located, using the same credentials that were used to push the image, or
+  [to the registry specified in the `COSIGN_REPOSITORY` environment variable](#key-generation-and-storage).
+  This flow must happen after the image has been pushed to the registry.
+
+- If `cosign` fails to sign an image, the build should fail and output an error
+  message in the build log, so the operator can troubleshoot the issue. The
+  errors should also appear in any other places where error messages are
+  presented.
+
+- Whenever `kpack` signs an image using `cosign`, it should add these
+  annotations:
+  - Build number.
+  - Build time.
+
+- Signing with `cosign` should not affect any configurations that enable signing
+  using the Notary v1 mechanism. These two signing mechanisms should be able to
+  coexist.
+
+### Key generation and storage
+
+`cosign` can use file-based keys, key management systems or hardware keys to
+ingest private keys, as well as a keyless flow that requires integration with
+[Rekor](https://github.com/sigstore/rekor).
+
+`kpack` should not generate the keys used for signing and verification. The user
+should pass them in using one of the mechanisms supported by `cosign`.
+In this RFC, we suggest that the first mechanism implemented be the use of
+Kubernetes `Secret`s for sending the relevant data into the build pod, by means
+of attaching secrets to the service account used by `kpack`.
+These secrets should contain private key and private key passphrase.
+Optionally, the user can add a repository URL to the secret, to address the use
+case when pushing the signatures to a different registry is desired.
+
+`kpack` will be able to determine that the user wants to sign images using
+`cosign` by scanning the types of the secrets attached to the service account,
+which in this use case would be a custom type:
+```yaml
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: cosign-key
+type: kpack.io/cosign-credentials
+stringData:
+  cosign-privatekey: <cosign-private-key>
+  passphrase: <cosign-private-key-passphrase>
+  # Optional repository URL, to be used as the environment variable
+  # COSIGN_REPOSITORY. If not set, pushes the signature collocated with
+  # the image.
+  cosign-repository-url: <repository-url>
```

Comment:
I think one more thing we might need to account for is the cosign variable `COSIGN_DOCKER_MEDIA_TYPES`
See https://github.com/sigstore/cosign#registry-support

There are some registries that work only with the variable set and some that work only without the variable. This should be exposed in the config somehow.

cc: @dlorenc who might know more about compat issues with registries.

Reply:
Yep, that one would be nice to have but probably not critical. It's rare to see issues and we don't have a perfect understanding across registries unfortunately.

Reply:
I have sadly faced issues without that flag in certain private registries. It would be great if this was included.

Reply:
Is there any risk in configuring this on the secret directly?

Reply:
If we do add this option I think it would make the most sense to add it to the secret directly since that's where the repository URL is.  The option is there to support a registry so the option should be located alongside the registry location you want to use.

One risk I see to putting this option in the secret is now that the secret has fields in it that are not really a secret; it's polluting a secret with non-secret things.  I'm not sure if this is idiomatic Kubernetes or not.

Reply:
I added this field to the proposal as an annotation, as discussed yesterday. 🙂